### PR TITLE
Persistent directories for project operations

### PR DESCRIPTION
**Pull request by johnsonr**
Uses persistent directories for project operations. This is useful for testing and debugging but also offers potential performance improvement.

### CODE REVIEW

File: src/project/git/StableDirectoryManager.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+import * as os from "os";
+import { CloneDirectoryInfo, CloneOptions, DirectoryManager } from "./DirectoryManager";
+
+export interface StableDirectoryManagerOpts {
+
+    /**
+     * Maximum number of directories to use
+     * @type {number}
+     */
+    maxDirectories?: number;
```
Comment (jessitron) [inline src/project/git/StableDirectoryManager.ts]: if this isn't implemented we should at least put in a comment saying so

File: src/project/git/StableDirectoryManager.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    /**
+     * Return undefined if not found
+     */
+    private existingDirectoryFor(user: string, repo: string, branch: string,
+                                 opts: CloneOptions): Promise<CloneDirectoryInfo> {
+        const expectedPath = this.pathFor(user, repo);
+        return fs.pathExists(expectedPath)
+            .then(exists => {
+                if (exists) {
+                    logger.info("%s directories used of %s: Reusing path [%s]",
```
Comment (jessitron) [inline src/project/git/StableDirectoryManager.ts]: wny would we care about that max on existing directories?

File: src/project/git/StableDirectoryManager.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                logger.info("%s directories used of %s: Returning new path [%s]",
+                    this.directoriesUsed, this.opts.maxDirectories, path);
+                return {
+                    path,
+                    type: "parent-directory" as any,
+                };
+            });
+    }
+
+    private pathFor(user: string, repo: string): string {
```
Comment (jessitron) [inline src/project/git/StableDirectoryManager.ts]: all this "user" is really "owner" right?

**Comment by jessitron** [comment]
@cdupuis Rod wanted you to look at this too. It makes the default checkout behavior be persistent.
This could be enhanced to do fetch instead of clone, which would be faster. It's also handy for testing; you can check out what the automation did after shutting down the client.

@johnsonr right now as this is, what is the behavior when we go to clone a repo, and the directory is already there?

File: src/project/git/StableDirectoryManager.ts (hunk 2 above)
Comment (johnsonr) [reply src/project/git/StableDirectoryManager.ts]: It's just a useful piece of log info

File: src/project/git/StableDirectoryManager.ts (hunk 3 above)
Comment (johnsonr) [reply src/project/git/StableDirectoryManager.ts]: Yeah

**Comment by johnsonr** [comment]
@jessitron Right now if the directory is already there it would be reused if the reuse flag is set in options. `GitCommandGitProject` needs to be modified to handle that correctly, so that flag cannot be used yet.

File: src/project/git/StableDirectoryManager.ts [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+    private directoriesUsed = 0;
+
+    constructor(pOpts: StableDirectoryManagerOpts) {
+        this.opts = {
+            ...DefaultOpts,
+            ...pOpts,
+        };
+
+        logger.info("Setting up StableDirectoryManager with options %s", JSON.stringify(this.opts));
```
Comment (cdupuis) [inline src/project/git/StableDirectoryManager.ts]: debug? also have you seen `logger.info("Setting up StableDirectoryManager with options %j"`? see https://nodejs.org/api/util.html#util_util_format_format_args. I only found about this recently.

File: src/project/git/StableDirectoryManager.ts [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+        if (this.opts.cleanOnExit === true) {
+            logger.info("Registering shutdown hook to delete [%s]", this.opts.baseDir);
```
Comment (cdupuis) [inline src/project/git/StableDirectoryManager.ts]: debug?

File: src/project/git/StableDirectoryManager.ts [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            registerShutdownHook(() => {
+                logger.info("Cleaning up temporary directories under [%s]", this.opts.baseDir);
```
Comment (cdupuis) [inline src/project/git/StableDirectoryManager.ts]: info ok here.

File: src/project/git/StableDirectoryManager.ts (hunk 2 above)
Comment (cdupuis) [reply src/project/git/StableDirectoryManager.ts]: debug

File: src/project/git/StableDirectoryManager.ts [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+                }
+            });
+    }
+
+    private freshDirectoryFor(user: string, repo: string, branch: string,
+                              opts: CloneOptions): Promise<CloneDirectoryInfo> {
+        return this.createDir(user, repo)
+            .then(path => {
+                this.directoriesUsed++;
+                logger.info("%s directories used of %s: Returning new path [%s]",
```
Comment (cdupuis) [inline src/project/git/StableDirectoryManager.ts]: debug

File: src/project/git/StableDirectoryManager.ts [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+/**
+ * Default Atomist working directory
+ * @type {string}
+ */
+const AtomistWorkingDirectory = ".atomist-working";
+
+export const PersistentDirectoryManager = new StableDirectoryManager({
+    baseDir: os.homedir() + "/" + AtomistWorkingDirectory,
+    cleanOnExit: false,
```
Comment (cdupuis) [inline src/project/git/StableDirectoryManager.ts]: `${os.homedir()}/${AtomistWorkingDirectory}`

File: src/project/git/StableDirectoryManager.ts (hunk 4 above)
Comment (johnsonr) [reply src/project/git/StableDirectoryManager.ts]: Thanks!

**Pull request merged**